functorch/functorch/_src/aot_autograd.py

```diff
@@ -59,6 +60,37 @@ def preserve_rng_state():
             torch.cuda.set_rng_state(cuda_rng_state)
 
 
+# Set up hooks so that during backward the fx's stack_trace is properly set
+def setup_stacktrace_preservation_hooks(roots: List):
+    def iter_graph(roots):
+        if not roots:
+            return
+        seen = set(roots)
+        q = collections.deque()
+        for node in roots:
+            q.append(node)
+
+        while q:
+            node = q.popleft()
+            for fn, _idx in node.next_functions:
+                if fn in seen or fn is None:
+                    continue
+                seen.add(fn)
+                q.append(fn)
+
+            yield node
+
+    def get_prehook(metadata_):
+        def prehook(grad_output):
+            stack = metadata_.get("traceback_", None)
+            fx_traceback.set_stack_trace(stack)
+
+        return prehook
+
+    for node in iter_graph(roots):
+        node.register_prehook(get_prehook(node.metadata))
```

Comment:
Aren't you missing the post hook here to remove the trace once that node is done executing?

Reply:
Yes, we can do that, but I am not sure if that's exactly necessary... 
If we have the post hook, here will be a access pattern... 

stack trace is A
For node 1: 
Override stack trace to B in prehook 
Override stack trace to A in post hook
For node 2: 
Override stack trace to C in prehook 
Override stack trace to A in post hook
.... 

Reverting  the stack trace in post hook for every node seems a bit excessive... 
and I am not sure how useful the initial stack trace will be anyway. In most of the case, I would guess the initial stack trace (A here) will be just empty.

Reply:
I think there are two differences:
- For calls coming from the engine (most notably addition within the input buffer), these will be attributed (wrongfully?) to the previous op. It might even be better to have a special stack trace for these to explain where they come from? (code that re-uses the same variable multiple times in user code that triggers an add in the backward).
- At the end of the backward, you will never clear back to the original stack?

Reply:
addressed.